common/src/chain/gen_block.rs

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) 2022 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://spdx.org/licenses/MIT
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Generalized block, or [GenBlock]
+
+use super::{Block, Genesis};
+use crate::primitives::Id;
+
+/// Generalized block that's either [genesis::Genesis] or [block::Block].
+///
+/// Does not contain any data, only used as `Id<GenBlock>` to signify given Id can refer to either
+/// genesis or proper block.
+#[derive(Eq, PartialEq, Clone, Debug)]
+pub enum GenBlock {}
+
+impl From<Id<Block>> for Id<GenBlock> {
+    fn from(id: Id<Block>) -> Id<GenBlock> {
+        Id::new(id.get())
+    }
+}
+
+impl From<Id<Genesis>> for Id<GenBlock> {
+    fn from(id: Id<Genesis>) -> Id<GenBlock> {
+        Id::new(id.get())
+    }
+}
+
+impl PartialEq<Id<Block>> for Id<GenBlock> {
+    fn eq(&self, other: &Id<Block>) -> bool {
+        self.get() == other.get()
+    }
+}
+
+impl PartialEq<Id<Genesis>> for Id<GenBlock> {
+    fn eq(&self, other: &Id<Genesis>) -> bool {
+        self.get() == other.get()
+    }
+}
+
+impl Id<GenBlock> {
+    /// Figure out if this [Id] refers to a [Genesis] or a proper [Block].
+    pub fn classify(&self, c: &crate::chain::config::ChainConfig) -> GenBlockId {
+        if self.get() == c.genesis_block_id().get() {
+            GenBlockId::Genesis(Id::new(self.get()))
+        } else {
+            GenBlockId::Block(Id::new(self.get()))
+        }
+    }
+}
+
+/// Classified generalized block
+pub enum GenBlockId {
+    Genesis(Id<Genesis>),
+    Block(Id<Block>),
+}
+
+impl GenBlockId {
+    pub fn is_genesis(&self) -> bool {
+        matches!(self, GenBlockId::Genesis(_))
+    }
+
+    pub fn block_id(self) -> Option<Id<Block>> {
```

Comment:
This is confusing. Why does Genesis return None? At least the function name is not appropriate. Something like `block_id_if_not_genesis` is more appropriate. I know it's a mouthful, but the current name is... confusing.

Reply:
Ok. The idea was to turn it into an option, so you get access to all the goodies `Option` has to offer. Regarding the name maybe `proper_block_id`? I already use the "proper block" terminology when I want to be explicit about it in some places I think.

Reply:
`chain_block_id()` should be OK, I'd say.

Reply:
Done